### PR TITLE
[codex] Add mirror geometry infrastructure

### PR DESCRIPTION
**Pull request by rogeriojorge**
## Summary

This is the long-lived draft PR for the mirror-geometry, differentiability, free-boundary planning, and toroidal stellarator-mirror hybrid lanes described in `plan_mirror.md`. It should stay draft pending user review; the square-coil stellarator-mirror hybrid lane is currently reopened for strict per-component `FTOL=1e-12` free-boundary evidence.

`plan_mirror.md` is the source of truth for detailed evidence, benchmark numbers, plot paths, and lane percentages. Current detailed implementation logs run through section 224. The newest implementation and validation tranche reopens the square-coil strict `FTOL=1e-12` profiling lane, adds live VMEC2000 sidecar progress reports, lets the profile summarizer read active VMEC2000 `threed1` runs, documents current mirror beta-validation references, and records the active `MPOL=7, NTOR=28` VMEC2000 near-floor profile while queued `8,32` and direct-coil GPU profiles continue.

## Current contents

- Experimental `vmec_jax.mirror` domain package for open-ended fixed-boundary mirror coordinates `(s, theta, xi)`.
- Chebyshev/Fourier mirror grids, geometry, field, energy, residual, `fsq`, normalized-force, active-force diagnostics, residual-Newton tooling, VMEC-like preconditioners, dense/block reference solves, and matrix-free diagnostics.
- Mirror-native `mout_*.nc` I/O plus `vmec --plot mout_*.nc` support with horizontal-z geometry, cross sections, `|B|`, field arrows, cap-to-cap field lines, residual history, radial beta, pitch, magnetic-well proxy, and mirror-Boozer-like plots.
- Analytic two-coil benchmark utilities with on-axis and low-radius circular-loop Biot-Savart comparisons, finite-current pitch examples, open-field cap-to-cap pitch diagnostics, and nonblank field-line plot coverage.
- Reduced mirror differentiability gates: JAX residuals/Jacobians, dense and matrix-free linear solves, Hessian-vector products, matrix-free CG, explicit forward/adjoint wrappers, custom VJP cached-state wrappers, finite-difference perturbed-root checks, and benchmark/example coverage.
- Free-boundary circular-coil bridge with ESSOS-compatible circular-loop parameters, beta-case metadata, normalized LCFS residual vector, combined equilibrium-plus-LCFS residual assembly, line-searched boundary LS steps, adaptive ridge diagnostics, realized fixed-boundary trials, guarded multi-step coupled LS loop, realized retry/downranking, frozen-LS accept tolerance, reduced residual-vector inner-solve fallback, adaptive polynomial-degree candidates, and schema `mirror_free_boundary_circular_coil_beta_scan` version `0.13`.
- Straight-axis open-ended stellarator-mirror hybrid fixture and toroidal stellarator-mirror hybrid VMEC-compatible boundary fixture, examples, convergence runner, target presets, aggregation support, plots, and local/office VMEC2000 parity evidence.
- Focused mirror and toroidal-hybrid tests plus docs under `docs/mirror/` and `examples/mirror/README.md`, with the mirror namespace single-sourced from `vmec_jax.mirror.api`.

## Current solver status

- The ordinary toroidal VMEC/JAX fixed-boundary path is used for the toroidal hybrid fixture, so preconditioning, residual normalization, WOUT writing, and profile helpers stay aligned with regular toroidal geometry.
- The toroidal hybrid target ladder has office GPU/VMEC2000 evidence over all six named target rows. The follow-up strict-closure run converged all six rows at `ftol=1e-8` with `max_iter = niter = 160` in 124-134 iterations. The square-coil free-boundary hybrid lane is now being profiled separately at per-component `FTOL=1e-12`; VMEC2000 generated-`mgrid` is the current robustness reference.
- The open-ended mirror residual-Newton lane has correctness-reference dense/block solves and diagnostic scalable Krylov paths, but the moderate finite-current row remains lambda dominated and is not yet a production tight-convergence claim.
- The reduced mirror differentiability lane has tested source, pressure, current, flux, and polynomial-boundary cached-state implicit derivative gates. These are reduced axisymmetric gates, not a final promoted full-equilibrium API.
- The free-boundary circular-coil mirror lane has moved beyond scalar pilot scaffolding. The current root example is still a host-side diagnostic coupled loop, not the final differentiable nonlinear free-boundary solver API, but M217 gives local and office target-merit evidence below `0.1` across the tested circular-coil beta/resolution ladder.
- The fast CLI/example path may use host-side NumPy/SciPy/Matplotlib for runtime and memory efficiency; differentiable research APIs should avoid differentiating through long host-side optimizer loops.

## Checklist

- [x] M0-M9: mirror package, kernels, fixed-boundary solvers, I/O/plotting, benchmarks, pitch diagnostics
- [x] M10/M11 status cleanup and docs synchronization through current implemented status
- [x] Reduced mirror residual/Jacobian, dense and matrix-free implicit derivative gates
- [x] Root examples for fixed-boundary diagnostics, implicit sensitivities, implicit parameter gradients, finite-current pitch, two-coil benchmarks, toroidal hybrid convergence, and circular-coil free-boundary workflows
- [x] Free-boundary circular-coil schema hardening with compact `summary_rows`
- [x] Realized coupled LCFS pilot candidate scoring with plotted 1%, 3%, and 10% beta evidence
- [x] Public normalized LCFS residual vector and combined equilibrium-plus-LCFS residual assembly helper
- [x] Line-searched free-boundary LS boundary step helper and realized fixed-boundary coupled trial
- [x] Guarded multi-step coupled LS loop for mirror free-boundary solves
- [x] Reduced residual-vector nonlinear LS solve loop for compact free-boundary prototypes
- [x] Reduced free-boundary LS rank/nullity/conditioning and selected-derivative-mode diagnostics
- [x] Adaptive ridge-candidate selection for reduced and combined free-boundary LS helpers
- [x] Realized retry/downranking, configurable frozen-LS accept tolerance, and inner-solve fallback for circular-coil coupled LS trials
- [x] Adaptive per-beta free-boundary polynomial-degree fallback with office resolution/runtime ladder evidence
- [x] Named toroidal hybrid target-resolution ladder preset, target case filtering, split-campaign aggregation, and office VMEC2000 parity/strict-closure evidence
- [x] Mirror-Boozer-like diagnostics promoted through plot, export, public API, and example JSON summary paths
- [x] Mirror readiness matrix documents supported, prototype, diagnostic, and deferred lanes through M219
- [x] Final explicit review-gate audit and final CI review for the prior closure point (PR remains draft pending user review)
- [ ] M220 strict square-coil `FTOL=1e-12` production evidence is active: VMEC2000 `7,28`/`8,32` and direct-coil GPU/JIT profiles are still running/queued

## Latest validation

Current M224 strict square-coil profiling gates:

```bash
venv/bin/python -m pytest -q tests/test_profile_square_coil_free_boundary.py
venv/bin/python -m pytest -q tests/test_summarize_square_coil_profiles.py
venv/bin/python -m pytest -q tests/test_toroidal_hybrid.py tests/test_profile_square_coil_free_boundary.py tests/test_summarize_square_coil_profiles.py tests/test_free_boundary_acceleration.py
venv/bin/python -m py_compile tools/diagnostics/profile_square_coil_free_boundary.py
venv/bin/python -m py_compile tools/diagnostics/summarize_square_coil_profiles.py
git diff --check
```

Results:

- Focused profile tests passed: `11 passed`.
- Focused summary tests passed: `3 passed`.
- Relevant square-hybrid/free-boundary suite passed after the live-profile changes: `56 passed, 2 warnings`.
- Active VMEC2000 `MPOL=7, NTOR=28, NZETA=64` copied-`threed1` summary reached visible final-stage iteration `14200` with summed residual about `2.38e-11`, max component about `1.07e-11`, best visible sum about `2.38e-11`, and `vacuum_grid_exceeded_count = 0`; strict per-component `1e-12` is not yet met.
- Queued runs remain active on `office`: VMEC2000 `MPOL=8, NTOR=32, NZETA=72`; direct-coil GPU/JIT `MPOL=6, NTOR=23` Anderson A/B; and direct-coil GPU/JIT `MPOL=7, NTOR=28` auto-mixer.

Most recent local gates for M217-M219:

```bash
python -m pytest tests/mirror/test_mirror_examples.py::test_root_free_boundary_circular_coils_summary_reports_converged_free_boundary_statuses tests/mirror/test_mirror_examples.py::test_root_free_boundary_circular_coils_polynomial_degree_candidate_helpers tests/mirror/test_mirror_examples.py::test_root_free_boundary_circular_coils_ls_boundary_step_reports_reduction tests/mirror/test_mirror_examples.py::test_root_free_boundary_circular_coils_high_order_ls_boundary_step_rejects_invalid_trials tests/mirror/test_mirror_examples.py::test_root_free_boundary_circular_coils_ls_boundary_coupled_trial_reports_realized_solve tests/mirror/test_mirror_examples.py::test_root_free_boundary_circular_coils_ls_boundary_coupled_loop_reports_guarded_steps tests/mirror/test_mirror_examples.py::test_root_free_boundary_circular_coils_ls_boundary_coupled_loop_reports_inner_solve_rows tests/mirror/test_mirror_examples.py::test_root_free_boundary_circular_coils_coupled_loop_reports_target_merit_convergence -q
python examples/mirror_free_boundary_circular_coils.py --outdir results/mirror/free_boundary_circular_coils_m217_candidate_schema_target010 --betas 1,3,10 --ntheta 8 --nxi 11 --n-segments 64 --run-fixed-boundary-baseline --baseline-maxiter 5 --run-ls-boundary-coupled-loop --ls-boundary-coupled-loop-steps 12 --ls-boundary-coupled-loop-target-merit 0.1 --ls-boundary-coupled-loop-fsq-growth-limit 1.5 --ls-boundary-max-relative-step 0.05 --ls-boundary-inner-solve-steps 2 --ls-boundary-polynomial-degree-candidates 4,8 --no-plots
python -m pytest tests/mirror/test_mirror_examples.py -q
python -m ruff check examples/mirror_free_boundary_circular_coils.py tests/mirror/test_mirror_examples.py
python -m ruff format --check examples/mirror_free_boundary_circular_coils.py tests/mirror/test_mirror_examples.py
python -m sphinx -W -j auto -b html docs docs/_build/html
git diff --check
```

Results:

- Local target probe with schema `0.13` and candidates `[4,8]` converged all default circular-coil beta rows below `0.1`: final merits were `0.06384411952516199` for beta 1%, and `0.04728522927730968` for beta 3% and 10%; stop reasons were `{"None": 24, "target_merit": 3}` with 27/27 loop rows accepted.
- Focused free-boundary tests passed: `8 passed in 65.68s`.
- Full root mirror example tests passed: `39 passed in 238.23s`.
- Full local mirror suite passed: `253 passed, 1 skipped in 334.42s`.
- Ruff check, Ruff format check, Sphinx with warnings as errors, plan section-order check, stale public-wording search, and whitespace check passed.
- GitHub CI at final head `55d59e01` completed successfully; `Physics Full (manual/nightly)` was skipped as expected, and Codecov project/patch checks passed.

Office M217 ladder with `--ls-boundary-polynomial-degree-candidates 4,8`, `max_steps=16`, and `inner_solve_steps=2`:

| ns | nxi | status | selected degrees by beta | final merits by beta | rows accepted/total | time | max RSS |
| ---: | ---: | --- | --- | --- | ---: | ---: | ---: |
| 7 | 11 | converged | `4, 4, 4` | `0.06384412004281963`, `0.047285219537912024`, `0.047285219537912024` | `27/27` | `2:31.00` | `1167560 KiB` |
| 9 | 15 | converged | `8, 4, 4` | `0.05857709842407309`, `0.09227264626028128`, `0.09227264626028128` | `15/15` | `2:14.10` | `1167984 KiB` |
| 11 | 17 | converged | `4, 4, 4` | `0.0632243695944156`, `0.05918247235266881`, `0.05918247235266881` | `37/37` | `4:31.12` | `1169604 KiB` |

At `(ns,nxi)=(9,15)`, beta 1% first tried degree 4, missed the target, then selected degree 8 and reached final merit `0.05857709842407309`. Degree 8 was not used where degree 4 already met the target.

## Notes for reviewers

- This PR is intentionally draft and broad. The detailed audit trail is in `plan_mirror.md`, and the current support matrix is in `docs/mirror/readiness.rst`.
- Generated result trees stay ignored under `results/`; only source, docs, tests, and plan logs are committed.
- The free-boundary lane remains labeled diagnostic/prototype until boundary parameters are promoted into a guarded differentiable nonlinear coupled residual solve. The current schema `0.13` captures inner-solve rows, candidate sources, realized retry rows, selected polynomial degree, per-degree attempts, and target-merit convergence evidence for the circular-coil diagnostic workflow.

---

### Latest update: M223 square-profile shape metadata

- Added explicit `--side-power` and `--corner-power` controls to the square-coil free-boundary profiler.
- Profile JSON and the summary table now record those shape powers beside `MPOL`, `NTOR`, and `NZETA`, preventing strict-resolution sweeps from being mixed with shape-smoothing sweeps.
- Focused verification: `tests/test_profile_square_coil_free_boundary.py` and `tests/test_summarize_square_coil_profiles.py` pass (`17 passed`, one existing JAX deprecation warning).
- Current strict-profile decision point remains numerical: active VMEC2000/direct long runs are still pending final rows; first-order-shape reruns are queued/ready before deeper spline-control work.

---

### Latest update: M226 production boundary-projection gate

- Added optional `--max-boundary-projection-error` to the square-coil backend profiler so production `MPOL`/`NTOR`/`NZETA` sweeps can fail before long backend solves when the Fourier boundary projection is too inaccurate.
- Successful profile JSON and summary rows now include `max_boundary_projection_error` next to observed `boundary_proj_max`.
- Plan log was cleaned up so M220-M226 are linear at the tail of `plan_mirror.md`.
- Focused verification: profile + summary tests pass (`18 passed`, one existing JAX deprecation warning), py-compile and whitespace checks pass.

---

### Latest update: M227 accepted-LCFS motion diagnostics

- Added accepted-boundary motion metrics to square-coil free-boundary profile JSON and summary tables: boundary_coeff_delta_* and boundary_sample_displacement_*.
- The profiler converts accepted VMEC internal edge coefficients back to physical Fourier coefficients before comparing them with the input boundary, and also measures sampled LCFS displacement on the solver angular grid.
- Updated docs and plan_mirror.md; focused verification passes: profile + summary tests (19 passed, one existing JAX deprecation warning), py-compile, and whitespace checks.

---

### Latest update: M228 VMEC2000 startup-phase reporting

- VMEC2000 partial profile payloads and summaries now include progress_phase, force_rows_started, and threed1 file status fields.
- Active rows with threed1 output but no parsed force table are labelled startup_or_pre_iteration_output instead of appearing as blank residual comparisons.
- Focused verification passes: profile + summary tests (21 passed, one existing JAX deprecation warning), py-compile, and whitespace checks.

---

### Latest update: M229 first-order spline projection floor

- Rechecked the current first-order square-axis spline target with the public projection diagnostic. Fixed MPOL=5 reaches the measured projection floor by NTOR=28: about 3.48e-12 max component error.
- Mixed high-mode rows show the same floor: 7,28 and 8,32 are both about 3.46e-12, so 8,32 is a useful robustness/performance probe but not a clear boundary-projection improvement.
- Active convergence docs now separate boundary projection closure from nonlinear/free-boundary solve-control evidence.

---

### Latest update: M230 queued VMEC2000 5,28 row

- Queued a VMEC2000 generated-mgrid profile for the first-order spline target at MPOL=5, NTOR=28, NZETA=64 with a 5e-12 boundary-projection gate.
- The corrected launcher waits behind the existing smooth 6,23 launcher and avoids concurrent VMEC2000 oversubscription. Aborted early launcher attempts and their orphaned xvmec process were cleaned up.
- Current active process set is the intended direct 6,23 Anderson run, VMEC2000 8,32 run, queued direct 7,28, queued VMEC2000 smooth 6,23, and queued VMEC2000 first-order 5,28.

### CODE REVIEW

**Comment by rogeriojorge** [comment]
GitHub cannot reopen this PR because its head branch was deleted and recreated after the main-history rewrite (API validation error). Clean draft successor: #22. The pre-rewrite head remains archived locally at e4a7f05d.
